docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+- HAZARD: Atom (and selector) key names need to be unique across the application. It can be easy to create an atom and selector with the same key.
+
+## useRecoilValue, useSetRecoilValue, useRecoilState hooks
+- PRO: The recoil state hooks cause rendering just like useState.
+- PRO: useRecoilValue and useSetRecoil value clearly denote the read-only and write-only scenarios.
+- PRO: No magic inspecting dot-notation usage like with MobX.  
+- HAZARD: Very easy to accidentally put a recoil hook within a condition. They need to be used as a block at the start of a component just like React hooks.
+
+## useRecoilCallback hook
+- PRO: Allows for creating methods that always get latest state even when called outside the context of a React component render. 
+- CON: It can be a bit of extra effort to get the Typescript types worked out for complex callbacks.
+- UNKNOWN: Unclear when to use the dependencies parameter vs. the getPromise to get latest state
+- UNKNOWN: Unclear when to use the synchronous loader vs the asynchronous getPromise.- UNKNOWN:? The callback documentation indicates is still possible this hook will change. This is likely due to the experimental portions relating to JS async threading.
+- HAZARD: Using getPromise within a callback requires the callback method be async, the return value be a promise, and usage outside UI callbacks use await.
+- HAZARD: Using set within a callback that is called multiple times in succession or called from other callbacks requires using updator function to ensure latest state.  Otherwise arrays may not be updated with the previous call and will be overridden - losing data.
+
+## RecoilRoot component
+- PRO: Simple to understand. It leverages the React context API, but more efficiently tracking atoms and selectors as keys in a dictionary.
+- HAZARD: The React component rendering RecoilRoot cannot use recoil hooks.
+? I didn't try nesting any RecoilRoots so I don't know if that really works.
+
+# Cost/Impact of moving to Recoil
+- Restructuring application state from observable hierarchy to flat arrays is painful and expensive. A major refactoring is likely to introduce bugs if there is insufficient testing on the existing code.
+- Converting deep dependencies on the context API down the component hierarchy is expensive especially when there are few pure functional components.
+- Creating selectors is straightforward. Many existing render calculations can be replaced incrementally with selectors.
+- Updating components to leverage recoil hooks rather than props with React.useEffect or React.useState is straightforward. Data passed through context can be updated with an atom or selector and then context usage can be replace incrementally down the component hierarchy.
+- Updating an existing dispatcher pattern is straightforward.
+
+# Tips and Best Practices
+
+## Follow functional programming paradigms
+- Always make state data immutable: All types used with atoms should have read only fields. 
```

Comment:
Can this be enforced? (ex. `Object.freeze`)

Reply:
Type definitions can use `readonly` as a keyword on object fields. I don't know where that gets enforced (I only just learned about it myself), but there's a mechanism already.

Reply:
Alternatively we can use `immer` (we already have this dependency in the app)